iscservice/demo/util.c

```diff
@@ -0,0 +1,710 @@
+	}
+	m0_conf_diter_fini(&it);
+	m0_confc_close(&root->rt_obj);
+	m0util_spiel_destroy(&spiel_inst);
+
+	return 0;
+}
+
+int m0util_isc_nxt_svc_get(struct m0_fid *svc_fid, struct m0_fid *nxt_fid,
+			   enum m0_conf_service_type s_type)
+{
+	struct m0_reqh_service_ctx *ctx;
+	struct m0_reqh             *reqh = &m0_instance->m0c_reqh;
+	struct m0_pools_common     *pc = reqh->rh_pools;
+	struct m0_fid               current_fid = *svc_fid;
+
+	m0_tl_for(pools_common_svc_ctx, &pc->pc_svc_ctxs, ctx) {
+		if (ctx->sc_type == s_type) {
+			if (m0_fid_eq(&current_fid, &M0_FID0)) {
+				*nxt_fid = ctx->sc_fid;
+				return 0;
+			} else if (m0_fid_eq(svc_fid, &ctx->sc_fid))
+				current_fid = M0_FID0;
+		}
+	} m0_tl_endfor;
+
+	*nxt_fid = M0_FID0;
+	return -ENOENT;
+}
+
+int m0util_isc_req_prepare(struct m0util_isc_req *req, struct m0_buf *args,
+			   const struct m0_fid *comp_fid,
+			   struct m0_layout_io_plop *iop, uint32_t reply_len)
+{
+	struct m0_rpc_session *sess = iop->iop_session;
+	struct m0_fop_isc  *fop_isc = &req->cir_isc_fop;
+	struct m0_fop      *arg_fop = &req->cir_fop;
+	int                 rc;
+
+	req->cir_plop = &iop->iop_base;
+	fop_isc->fi_comp_id = *comp_fid;
+	m0_rpc_at_init(&fop_isc->fi_args);
+	rc = m0_rpc_at_add(&fop_isc->fi_args, args, sess->s_conn);
+	if (rc != 0) {
+		m0_rpc_at_fini(&fop_isc->fi_args);
+		fprintf(stderr, "error! m0_rpc_at_add() failed with %d\n", rc);
+		return rc;
+	}
+	/* Initialise the reply RPC AT buffer to be received.*/
+	m0_rpc_at_init(&fop_isc->fi_ret);
+	rc = m0_rpc_at_recv(&fop_isc->fi_ret, sess->s_conn, reply_len, false);
+	if (rc != 0) {
+		m0_rpc_at_fini(&fop_isc->fi_args);
+		m0_rpc_at_fini(&fop_isc->fi_ret);
+		fprintf(stderr, "error! m0_rpc_at_recv() failed with %d\n", rc);
+		return rc;
+	}
+	m0_fop_init(arg_fop, &m0_fop_isc_fopt, fop_isc, m0_fop_release);
+	req->cir_rpc_sess = sess;
+
+	return rc;
+}
+
+void isc_req_replied(struct m0_rpc_item *item)
+{
+	int                    rc;
+	struct m0_fop         *fop = M0_AMB(fop, item, f_item);
+	struct m0_fop         *reply_fop;
+	struct m0_fop_isc_rep *isc_reply;
+	struct m0util_isc_req *req = M0_AMB(req, fop, cir_fop);
+	const char *addr = m0_rpc_conn_addr(req->cir_rpc_sess->s_conn);
+
+	if (item->ri_error != 0) {
+		req->cir_rc = item->ri_error;
+		fprintf(stderr,
+			"No reply from %s: rc=%d.\n", addr, item->ri_error);
+		goto err;
+	}
+	reply_fop = m0_rpc_item_to_fop(req->cir_fop.f_item.ri_reply);
+	isc_reply = (struct m0_fop_isc_rep *)m0_fop_data(reply_fop);
+	rc = req->cir_rc = isc_reply->fir_rc;
+	if (rc != 0) {
+		fprintf(stderr,
+			"Got error in reply from %s: rc=%d.\n", addr, rc);
+		if (rc == -ENOENT)
+			fprintf(stderr, "Was isc .so library is loaded?\n");
+		goto err;
+	}
+	rc = m0_rpc_at_rep_get(&req->cir_isc_fop.fi_ret, &isc_reply->fir_ret,
+			       &req->cir_result);
+	if (rc != 0)
+		fprintf(stderr,
+			"rpc_at_rep_get() from %s failed: rc=%d\n", addr, rc);
+ err:
+	m0_fop_put(&req->cir_fop);
+	m0_semaphore_up(req->cir_sem);
```

Comment:
Just call:
m0_semaphore_up(&isc_sem);

No need to embed this semaphore in req.

Reply:
Fixed.